### ISSUE TITLE
Disallowing assignment and passing by value.

### ISSUE THREAD

**Issue by xiaoxiangmoe** [eslint/eslint#12667]
**Please describe what the rule should do:**

Disallowing assignment and passing by value.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
if(exitCode = 0){
}
```
```js
const exitStatus = {
   noError: exitCode = 0,
}
```

Maybe we can consider a stricter version that strictly distinguishes statements with side effects.

Statements with side effects (such as `a=b`, `a+=b`, `a++`,`a--`,`delete a.b`) are prohibited from using as a value/expressions.

So what name should this rule be called? I encountered a naming problem.

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a strict version of [no-cond-assign](https://eslint.org/docs/rules/no-cond-assign) and [no-multi-assign](https://eslint.org/docs/rules/no-multi-assign). and it may also deprecating [yoda](https://eslint.org/docs/rules/yoda)

**Are you willing to submit a pull request to implement this rule?**
Yes

---

this idea comes from http://www.yinwang.org/blog-cn/2013/04/16/yoda-notation

translate: https://translate.google.com/translate?sl=zh-CN&tl=en&u=http%3A%2F%2Fwww.yinwang.org%2Fblog-cn%2F2013%2F04%2F16%2Fyoda-notation

**Comment by kaicataldo** [comment]
My only concern is that creating a new rule will create multiple rules that warn about the same thing. I think we either need to scope this proposal to the cases not currently covered or discuss deprecating `no-cond-assign` and `no-multi-assign` in favor of a new rule that catches all these cases (with configuration options to allow turning currently supported cases on and off individually).

**Comment by xiaoxiangmoe** [comment]
@kaicataldo Any suggestions for names and options? Should I start implementing it now?

**Comment by kaicataldo** [comment]
@xiaoxiangmoe We recommend waiting until the ESLint team comes to a consensus on this issue and we mark it as "accepted", so that people don't end up having throwaway work.

I'm honestly not sure what the best course of action is here, though I agree that warning on cases like the following one you listed above could be useful:
```js
const exitStatus = {
   noError: exitCode = 0,
}
```

**Comment by xiaoxiangmoe** [comment]
@kaicataldo Any update on this?

**Comment by kaicataldo** [comment]
As the bot above mentions, it doesn't look like we were able to drum up enough support for this proposal. ESLint is fully pluggable, though, and you can read about making your own custom rules [here](https://eslint.org/docs/developer-guide/working-with-plugins)! Feel free to stop by our [Gitter](https://gitter.im/eslint/eslint) if you need help getting this to work.
